### PR TITLE
Add the check for noqueue tun device

### PR DESCRIPTION
**Pull request by yalzhang**
When interface set without bandwidth, the default qdisc type should be
"noqueue". If the bandwidth is not set, the qdisc type should be "htb".

Signed-off-by: Yalan Zhang <yalzhang@redhat.com>

### CODE REVIEW

**Comment by yalzhang** [comment]
run the cases locally and pass
$ avocado run  --vt-type libvirt qos_option_bandwidth
JOB ID     : a2d03a24e97ce3f3b0822db5767c4dc7bd0c41dc
JOB LOG    : /root/avocado/job-results/job-2021-04-21T03.45-a2d03a2/job.log
 (1/4) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.positive_test: PASS (30.12 s)
 (2/4) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.by_update_device: PASS (33.52 s)
 (3/4) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_start_test: PASS (8.73 s)
 (4/4) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_remove_test: PASS (31.30 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 105.46 s
This is to auto the case RHEL-196264

**Comment by yalzhang** [comment]
some scenarios are included in existing case: virtual_network.iface_network.qos_test.qos_option_bandwidth.by_update_device
and one more auto case to cover one more scenario:
virtual_network.iface_network.qos_test.qos_option_bandwidth.qdisc_multiqueue

**Comment by yalzhang** [comment]
run locally pass:
 (13/46) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.positive_test: PASS (29.38 s)
 (14/46) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.by_update_device: PASS (33.21 s)
 (15/46) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.qdisc_multiqueue: PASS (28.12 s)
 (16/46) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_start_test: PASS (8.53 s)
 (17/46) type_specific.io-github-autotest-libvirt.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_remove_test: PASS (30.05 s)

**Comment by yalzhang** [comment]
depends on https://github.com/avocado-framework/avocado-vt/pull/3015

**Comment by yalzhang** [comment]
This it to cover RHEL-196264

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 1]
```diff
@@ -560,6 +563,19 @@ def run_guest_libvirt(session):
         if not utils_package.package_remove("libvirt*", session):
             test.error("Failed to remove libvirt packages on guest")
 
+    def check_qdisc(expect_type):
```
Comment (chloerh) [inline libvirt/tests/src/virtual_network/iface_network.py:566]: param `expect_type` 's missing from docstring

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 2]
```diff
@@ -758,8 +775,16 @@ def run_guest_libvirt(session):
             try:
                 if update_device:
                     updated_iface = modify_iface_xml(sync=False)
+                    if loop == 3:
+                        check_qdisc("noqueue")
+                    else:
+                        check_qdisc("htb")
                     virsh.update_device(vm_name, updated_iface.xml,
                                         ignore_status=False, debug=True)
+                    if loop == 3 or loop == 2:
```
Comment (chloerh) [inline libvirt/tests/src/virtual_network/iface_network.py:784]:
```suggestion
                    if loop in (2, 3):
```

**Comment by chloerh** [comment]
Please re-request me when you're ready.

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 3]
```diff
@@ -560,6 +563,19 @@ def run_guest_libvirt(session):
...
+    def check_qdisc(expect_type):
+        """
+        check the qdisc type of the tap device
+        """
+        if not vm.is_alive():
+            vm.start()
+        mac = vm_xml.VMXML.get_iface_dev(vm_name)[0]
+        iface_features = vm_xml.VMXML.get_iface_by_mac(vm_name, mac)
+        target_dev = iface_features.get('target', {}).get('dev')
+        outputs = process.run("ip l show %s" % target_dev).stdout_text
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_network/iface_network.py:575]: the default value of iface_features.target is {}, so do you need to test target_dev's value before executing the ip command?

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 4]
```diff
@@ -945,18 +970,20 @@ def _check_lease(loop):
...
-            ori_iface = params['original_iface']
-            logging.debug(ori_iface)
-            virsh.update_device(vm_name, ori_iface.xml,
-                                ignore_status=False, debug=True)
-
+            check_qdisc("htb")
+            inbound = outbound = '{"average":0}'
+            iface_dict = {"inbound": inbound, "outbound": outbound}
+            new_iface_xml = libvirt.modify_vm_iface(vm_name, "get_xml", iface_dict)
+            logging.debug(new_iface_xml)
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_network/iface_network.py:977]: iface xml will be logged if oper is "get_xml", so is this line necessary? 
https://github.com/avocado-framework/avocado-vt/blob/2fe4b2a8ee4a1b2ca0dc192bb5fea5d67d5c0402/virttest/utils_test/libvirt.py#L3588

Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: Thank you! Updated.

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 3 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated. Thank you!

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 1 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py:566]: Updated, Thank you!

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 2 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated, Thank you! Please help to review it again. @chloerh

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 3 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: @Yingshun  Please help to review it again. Thank you!

**Review by chloerh — APPROVED** [review]

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 5]
```diff
@@ -560,6 +563,22 @@ def run_guest_libvirt(session):
         if not utils_package.package_remove("libvirt*", session):
             test.error("Failed to remove libvirt packages on guest")
 
+    def check_qdisc(expect_type):
+        """
+        check the qdisc type of the tap device
+
+        param expect_type: the expected qdisc type for the interface, for example, noqueue, mq, htb or others.
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_network/iface_network.py:570]: Can we update this line to start with ':'?

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 6]
```diff
@@ -560,6 +563,22 @@ def run_guest_libvirt(session):
...
+        """
+        if not vm.is_alive():
+            vm.start()
+        mac = vm_xml.VMXML.get_iface_dev(vm_name)[0]
+        iface_features = vm_xml.VMXML.get_iface_by_mac(vm_name, mac)
+        target_dev = iface_features.get('target', {}).get('dev')
+        if target_dev:
+            outputs = process.run("ip l show %s" % target_dev).stdout_text
+        if not re.findall(expect_type, outputs):
+            test.fail("The qdisc type should be %s, but it is not, check %s" % (expect_type, outputs))
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_network/iface_network.py:580]: outputs will not be defined if target_dev is None, right? Should we raise an exception in this case?

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 5 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated, Thank you!

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 6 above)
Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: Add a check before the line, and update it to be:
        if not target_dev:
            test.skip("Failed to get the target dev name for qdisc test!")
        outputs = process.run("ip l show %s" % target_dev).stdout_text

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 7]
```diff
@@ -560,6 +563,23 @@ def run_guest_libvirt(session):
...
+
+        :param expect_type: the expected qdisc type for the interface, for example, noqueue, mq, htb or others.
+        """
+        if not vm.is_alive():
+            vm.start()
+        mac = vm_xml.VMXML.get_iface_dev(vm_name)[0]
+        iface_features = vm_xml.VMXML.get_iface_by_mac(vm_name, mac)
+        target_dev = iface_features.get('target', {}).get('dev')
+        if not target_dev:
+            test.skip("Failed to get the target dev name for qdisc test!")
```
Comment (Yingshun) [inline libvirt/tests/src/virtual_network/iface_network.py:578]: test.skip should be test.cancel or test.error, right?

Comment (yalzhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: updated, Thank you!

**Review by Yingshun — APPROVED** [review]

**Pull request merged**